### ISSUE TITLE
Detect expired (drive) token

### ISSUE THREAD

**Issue by abitrolly** [rclone/rclone#3073]
#### What is your current rclone version (output from `rclone version`)?

```
✗ rclone version
rclone v1.45
- os/arch: linux/amd64
- go version: go1.11.2
```

#### What problem are you are trying to solve?

Fix the workflow with expired Google Drive token. Which currently is this.
```
✗ rclone lsd gdrive:
2019/03/24 13:19:10 ERROR : : error listing: couldn't list directory: Get https://www.googleapis.com/drive/v3/files?alt=json&fields=files%28id%2Cname%2Csize%2Cmd5Checksum%2Ctrashed%2CmodifiedTime%2CcreatedTime%2CmimeType%2Cparents%2CwebViewLink%29%2CnextPageToken&pageSize=1000&prettyPrint=false&q=trashed%3Dfalse+and+%28%27root%27+in+parents%29: oauth2: cannot fetch token: 400 Bad Request
Response: {
  "error": "invalid_grant",
  "error_description": "Bad Request"
}
2019/03/24 13:19:10 Failed to lsd: couldn't list directory: Get https://www.googleapis.com/drive/v3/files?alt=json&fields=files%28id%2Cname%2Csize%2Cmd5Checksum%2Ctrashed%2CmodifiedTime%2CcreatedTime%2CmimeType%2Cparents%2CwebViewLink%29%2CnextPageToken&pageSize=1000&prettyPrint=false&q=trashed%3Dfalse+and+%28%27root%27+in+parents%29: oauth2: cannot fetch token: 400 Bad Request
Response: {
  "error": "invalid_grant",
  "error_description": "Bad Request"
}
```

```
✗ rclone config     
Current remotes:

Name                 Type
====                 ====
gdrive               drive

e) Edit existing remote
n) New remote
d) Delete remote
r) Rename remote
c) Copy remote
s) Set configuration password
q) Quit config
e/n/d/r/c/s/q> e
Choose a number from below, or type in an existing value
 1 > gdrive
remote> 1
--------------------
[gdrive]
type = drive
client_id = 
client_secret = 
scope = drive
root_folder_id = 
service_account_file = 
token = {"access_token":"ya29...fWGKZKvXw","token_type":"Bearer","refresh_token":"1/AcEkLseQBVuc6Ha_3MqCVPMgFbUGT1VSQgo6YZt8v30","expiry":"2018-06-30T09:56:28.688353217+07:00"}
--------------------
Edit remote
** See help for drive backend at: https://rclone.org/drive/ **

Value "client_id" = ""
Edit? (y/n)>
y) Yes
n) No
y/n> 
y/n> n
Value "client_secret" = ""
Edit? (y/n)>
y) Yes
n) No
y/n> n
Value "scope" = "drive"
Edit? (y/n)>
y) Yes
n) No
y/n> n
Value "root_folder_id" = ""
Edit? (y/n)>
y) Yes
n) No
y/n> n
Value "service_account_file" = ""
Edit? (y/n)>
y) Yes
n) No
y/n> n
Edit advanced config? (y/n)
y) Yes
n) No
y/n> n
--------------------
[gdrive]
type = drive
client_id = 
client_secret = 
scope = drive
root_folder_id = 
service_account_file = 
token = {"access_token":"ya29...fWGKZKvXw","token_type":"Bearer","refresh_token":"1/AcEkLseQBVuc6Ha_3MqCVPMgFbUGT1VSQgo6YZt8v30","expiry":"2018-06-30T09:56:28.688353217+07:00"}
--------------------
y) Yes this is OK
e) Edit this remote
d) Delete this remote
y/e/d> e
** See help for drive backend at: https://rclone.org/drive/ **

Value "client_id" = ""
Edit? (y/n)>
y) Yes
n) No
y/n> n
Value "client_secret" = ""
Edit? (y/n)>
y) Yes
n) No
y/n> n
Value "scope" = "drive"
Edit? (y/n)>
y) Yes
n) No
y/n> n
Value "root_folder_id" = ""
Edit? (y/n)>
y) Yes
n) No
y/n> n
Value "service_account_file" = ""
Edit? (y/n)>
y) Yes
n) No
y/n> n
Edit advanced config? (y/n)
y) Yes
n) No
y/n> y
Value "service_account_credentials" = ""
Edit? (y/n)>
y) Yes
n) No
y/n> y
Service Account Credentials JSON blob
Leave blank normally.
Needed only if you want use SA instead of interactive login.
Enter a string value. Press Enter for the default ("").
service_account_credentials> 
Value "team_drive" = ""
Edit? (y/n)>
y) Yes
n) No
y/n> n
Value "auth_owner_only" = ""
Edit? (y/n)>
y) Yes
n) No
y/n> n
Value "use_trash" = ""
Edit? (y/n)>
y) Yes
n) No
y/n> n
Value "skip_gdocs" = ""
Edit? (y/n)>
y) Yes
n) No
y/n> n
Value "shared_with_me" = ""
Edit? (y/n)>
y) Yes
n) No
y/n> n
Value "trashed_only" = ""
Edit? (y/n)>
y) Yes
n) No
y/n> n
Value "formats" = ""
Edit? (y/n)>
y) Yes
n) No
y/n> n
Value "export_formats" = ""
Edit? (y/n)>
y) Yes
n) No
y/n> n
Value "import_formats" = ""
Edit? (y/n)>
y) Yes
n) No
y/n> n
Value "allow_import_name_change" = ""
Edit? (y/n)>
y) Yes
n) No
y/n> n
Value "use_created_date" = ""
Edit? (y/n)>
y) Yes
n) No
y/n> n
Value "list_chunk" = ""
Edit? (y/n)>
y) Yes
n) No
y/n> n
Value "impersonate" = ""
Edit? (y/n)>
y) Yes
n) No
y/n> n
Value "alternate_export" = ""
Edit? (y/n)>
y) Yes
n) No
y/n> n
Value "upload_cutoff" = ""
Edit? (y/n)>
y) Yes
n) No
y/n> n
Value "chunk_size" = ""
Edit? (y/n)>
y) Yes
n) No
y/n> n
Value "acknowledge_abuse" = ""
Edit? (y/n)>
y) Yes
n) No
y/n> n
Value "keep_revision_forever" = ""
Edit? (y/n)>
y) Yes
n) No
y/n> n
Value "v2_download_min_size" = ""
Edit? (y/n)>
y) Yes
n) No
y/n> n
--------------------
[gdrive]
type = drive
client_id = 
client_secret = 
scope = drive
root_folder_id = 
service_account_file = 
token = {"access_token":"ya29...fWGKZKvXw","token_type":"Bearer","refresh_token":"1/AcEkLseQBVuc6Ha_3MqCVPMgFbUGT1VSQgo6YZt8v30","expiry":"2018-06-30T09:56:28.688353217+07:00"}
--------------------
y) Yes this is OK
e) Edit this remote
d) Delete this remote
y/e/d> e
** See help for drive backend at: https://rclone.org/drive/ **

Value "client_id" = ""
Edit? (y/n)>
y) Yes
n) No
y/n> n
Value "client_secret" = ""
Edit? (y/n)>
y) Yes
n) No
y/n> n
Value "scope" = "drive"
Edit? (y/n)>
y) Yes
n) No
y/n> n
Value "root_folder_id" = ""
Edit? (y/n)>
y) Yes
n) No
y/n> n
Value "service_account_file" = ""
Edit? (y/n)>
y) Yes
n) No
y/n> n
Edit advanced config? (y/n)
y) Yes
n) No
y/n> n
--------------------
[gdrive]
type = drive
client_id = 
client_secret = 
scope = drive
root_folder_id = 
service_account_file = 
token = {"access_token":"ya29...fWGKZKvXw","token_type":"Bearer","refresh_token":"1/AcEkLseQBVuc6Ha_3MqCVPMgFbUGT1VSQgo6YZt8v30","expiry":"2018-06-30T09:56:28.688353217+07:00"}
--------------------
y) Yes this is OK
e) Edit this remote
d) Delete this remote
y/e/d> y
Remote config
Already have a token - refresh?
y) Yes
n) No
y/n> y
Use auto config?
 * Say Y if not sure
 * Say N if you are working on a remote or headless machine or Y didn't work
y) Yes
n) No
y/n> y
If your browser doesn't open automatically go to the following link: http://127.0.0.1:53682/auth
Log in and authorize rclone for access
Waiting for code...
Got code
Configure this as a team drive?
y) Yes
n) No
y/n> n
Current remotes:

Name                 Type
====                 ====
gdrive               drive

e) Edit existing remote
n) New remote
d) Delete remote
r) Rename remote
c) Copy remote
s) Set configuration password
q) Quit config
e/n/d/r/c/s/q> q 
✗ rclone lsd gdrive:
          -1 2012-12-26 00:17:17        -1 .art
```

#### How do you think rclone should be changed to solve that?
```
✗ rclone lsd gdrive:
2019/03/24 13:19:10 ERROR : : error listing: token is expired
You GDrive token is expired. Refresh? (y/n)
```
or for non-interactive usage
```
✗ rclone lsd gdrive:
2019/03/24 13:19:10 ERROR : : error listing: token is expired

You Google Drive token is expired. Run `rclone refresh token gdrive:` to update.
```
Thanks to https://github.com/ncw/rclone/issues/1364#issuecomment-296514445 for the hint about token refresh.

**Comment by ncw** [comment]
Normally rclone uses the refresh token to refresh the credentials and everything works just fine.

I've never seen it go wrong but it does go wrong sometimes.  In fact I found [an article with all the ways it can go wrong](https://blog.timekit.io/google-oauth-invalid-grant-nightmare-and-how-to-fix-it-9f4efaf1da35).

So what I think you are asking for is that rclone detects this error

> cannot fetch token: 400 Bad Request
> Response: {
>   "error": "invalid_grant",
>   "error_description": "Bad Request"
> }

And translates it into something more user friendly.

This should definitely not be happening in normal operation though - can you replicate the problem reliably?  If so can you work out why it is happening?

**Comment by animosity22** [comment]
This looks to be stale. If you have a question, please just make a post on the forums and we can help out.
